### ISSUE TITLE
[feature] gfortran Compiler Support

### ISSUE THREAD

**Issue by dnwillia** [conan-io/conan#15430]
### What is your suggestion?

Hi! I'm working on a library that is built from Fortran source and redistributed as a Conan package.  I'm using gfortran for that, which is not supported 'out of the box', so here's the content I manually added to the `settings.yml`:
```
compiler:
    ...
    gfortran:
        version: ["4.1", "4.4", "4.5", "4.6", "4.7", "4.8", "4.9", "4.9.1",
                    "5", "5.1", "5.2", "5.3", "5.4", "5.5",
                    "6", "6.1", "6.2", "6.3", "6.4", "6.5",
                    "7", "7.1", "7.2", "7.3", "7.4", "7.5",
                    "8", "8.1", "8.2", "8.3", "8.4", "8.5",
                    "9", "9.1", "9.2", "9.3", "9.4", "9.5",
                    "10", "10.1", "10.2", "10.3", "10.4", "10.5"]
        integer_size: [4, 8, 16]
        real_size: [4, 8, 16]
        underscoring: ["default", "none", "second"]
        fortran_std: ["f95", "f2003", "f2008", "f2018", "f2023"]
```
The version numbers are aligned with the gcc numbering but it's not as far along as you can see.  The last four settings are selectable using compiler flags and impact library consumers. I can push up the changes in a PR if you are interested.  If I found the right spot looks like the defaults go [here](https://github.com/conan-io/conan/blob/2985accf41e9380ba5b12dd8b4392b904abb4c21/conans/client/conf/__init__.py#L15).

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @dnwillia 

Thanks for your suggestion and offer to help.

It would be good to know a bit better before proceeding how these settings might be used or if they are expected to be used somehow by built-in build system integrations like ``CMakeToolchain`` if that is the case. And also how they would play with other C++ packages.

I assume that this would be always necessary to manage with per-package settings, and define profiles that specifically define the fortran settings for some packages. Also I guess these are pure fortran packages without a C/C++ interface that would need the standard C/C++ settings

If they are only used as settings, but Conan doesn't need to be aware of them, it might be possible to just rely on ``settings_user.yml``, but I guess that is not the case, and the idea is that Conan can use the settings to do things automatically.

Also to move this forward it would be good to know about the binary compatibility of the:
```
        integer_size: [4, 8, 16]
        real_size: [4, 8, 16]
        underscoring: ["default", "none", "no_second"]
        fortran_std: ["f95", "f2003", "f2008", "f2018", "f2023"]
```
I guess the type sizes really generate incompatible binaries, but no idea about the underscoring, and if it could be defaulted (not specified) with ``underscoring: [null, "value1", "value2"]``

**Comment by dnwillia** [comment]
Regarding the built in CMake toolchain I'm not certain because I'm using SCons.  Fundamentally though, besides the compiler version, the additional settings correspond to compiler command line options you might also use with CMake.  Consuming C++ packages could be impacted if a C++ package is calling through to the Fortran subroutines and functions.  In that case as a consumer you may need to know the integer and real byte sizes so you can use the appropriate C type such as `float` for 4 byte Fortran integers or `double` for 8 byte.  That's the reason I included this information in the settings.

I have not looked at `settings_user.yml` before, so will do that.  If that's the normal way to handle this sort of thing to modify the defaults then that's fine too.  I'm still kind of learning Conan.

None of the settings impact the binary compatibility as long as any consuming Fortran code is using the same Fortran runtime, or, equivalently, compiler version.  If you are running a different compiler version then the ABIs would generally be incompatible, in the same way that they would be with different versions of g++ and libstdc++.

Regarding the underscoring it's a Fortran thing about how the external names are mangled, and are there for backwards compatibility with older Fortran codes compiled when the mangling strategies were different.  The default, which I called `default`, is just what gfortran does by default if you do nothing.  In that case all the external names have an underscore `_` character appended to the name.  If you activate the second underscore option then two underscores `__` are appended if the external name contains an underscore.  The different settings indicate what you have activated.  I got my OP wrong and have updated the second underscoring thing, it should just have said `second` not `no_second`.

In retrospect, I could probably leave the default setting out of the list to simplify things.  i.e. default is no setting for that property.

Here's the [compiler documentation](https://gcc.gnu.org/onlinedocs/gfortran/Code-Gen-Options.html#index-fno-underscoring) for that detail (if it helps).

**Comment by memsharded** [comment]
Ok, thanks for the feedback, I understand now a bit better.
I assume then your project contains in the SCons scripts the necessary assumptions to handle fortran code, and you might be passing in your ``conanfile.py`` recipe some info to your SCons scripts

Just to clarify, are you using Conan for just fortran code, to be consumed by other fortran packages, or is it just this package that is mostly used by C and C++ packages? You say you are distributing it, but if it uses custom settings, those settings also have to be distributed to your users, and depending who your users are this might not be the most convenient (for example, if trying to distribute to the world cause it is an open source project)

> None of the settings impact the binary compatibility as long as any consuming Fortran code is using the same Fortran runtime, or, equivalently, compiler version. If you are running a different compiler version then the ABIs would generally be incompatible, in the same way that they would be with different versions of g++ and libstdc++.

You mean that different fortran libraries can use different values for those settings, and they will happily link and work together without issues? If that is the case, it might be that you could want to implement a custom ``compatibility.py`` to implement that compatibility, if you don't want to force exact matching of your binaries. Or maybe drop it, and let each package define (hardcoded) which one are they using and just inform their potential consumers via ``cpp_info.defines`` preprocessor definitions or via some ``conf_info``. 

Also, for the same version of g++, different values of ``libcxx`` will be completely incompatible binaries, this is why ``compiler.libcxx`` doesn't have any provision in the ``compatibility.py``. The only assumption we have so far for the C++ binaries is compatibility for different ``cppstd`` standards.

> In retrospect, I could probably leave the default setting out of the list to simplify things. i.e. default is no setting for that property.

It is different for Conan allowing a setting to not be defined than being the default. If there is a default value for a compiler, for example, gcc 11 uses cppstd=14 by default, then the recommendation is to have to explicitly define it. It could only be undefined for example if some user is using only C packages and not C++, they could leave the ``compiler.cppstd`` undefined. But that doesn't mean that it will be the compiler default, but that it doesn't apply. This point is important for the ``package_id`` computation and future compatibility.

**Comment by samuel-emrys** [comment]
For tracking, I've made some previous attempts to introduce gfortran as a viable compiler/dependency for conan packages on cci. However, none of these have attempted to explicitly model package metadata as "fortran" packages:

* https://github.com/conan-io/conan-center-index/issues/15099
* https://github.com/conan-io/conan-center-index/pull/15128
* https://github.com/conan-io/conan/issues/13533

From a metadata perspective, I think "gfortran" as a compiler value makes sense for projects that are exclusively fortran projects. However, my experience (as a C++ developer) has been that it's more likely that a project depends on both C/C++ and fortran, which complicates things - how can we model the metadata for two different compilers per project? should the metadata be tracked at the library level instead of the package level?

**Comment by memsharded** [comment]
Yes, one of the problems of having ``gfortran`` as a compiler is the mixture of languages in the same package. In that sense, it would be better to define a new ``fortran_compiler`` setting category, so packages can do ``settings = "os", "arch", "compiler", "fortran_compiler"``, (assuming ``compiler`` is C/C++) or just ``settings = "os", "arch", "fortran_compiler"`` if they are pure fortran packages.

**Comment by samuel-emrys** [comment]
What are the situations in which the ABI of fortran binaries/archives needs to be tracked (aside from modelling dependencies between pure fortran libraries)? Is the risk here that if we have two dependencies that are both C++ and Fortran projects, that if they've used a different fortran compiler version that we'll have an ABI mismatch when both dependencies are static?

To date the bigger concern that I've had is that the version of libgfortran available on a system doesn't match the version that a binary has been compiled with (or doesn't exist, because it's not part of the default gcc toolchain on most linux systems)

**Comment by dnwillia** [comment]
> Ok, thanks for the feedback, I understand now a bit better. I assume then your project contains in the SCons scripts the necessary assumptions to handle fortran code, and you might be passing in your `conanfile.py` recipe some info to your SCons scripts

Yes, exactly.  I can manually build, independent from Conan, and then export the package or use the `build` method in the recipe.

> Just to clarify, are you using Conan for just fortran code, to be consumed by other fortran packages, or is it just this package that is mostly used by C and C++ packages? You say you are distributing it, but if it uses custom settings, those settings also have to be distributed to your users, and depending who your users are this might not be the most convenient (for example, if trying to distribute to the world cause it is an open source project)

It's a library build from Fortran source meant to be linked into other Fortran or C applications, but also C++ by extension via C right.  The project is internal so we've agreed on the settings that correspond to the specific compiler options, but you are right, for wider consumption in the outside world it would be important to have some consensus.

> You mean that different fortran libraries can use different values for those settings, and they will happily link and work together without issues? If that is the case, it might be that you could want to implement a custom `compatibility.py` to implement that compatibility, if you don't want to force exact matching of your binaries. Or maybe drop it, and let each package define (hardcoded) which one are they using and just inform their potential consumers via `cpp_info.defines` preprocessor definitions or via some `conf_info`.

Yes, that's right, consumers can use different values for the byte sizes than the library but as a consumer you need to know what the library you are consuming **has** used.  Fortran has this concept of generic types for INTEGER and REAL and you need to know if the library has forced the byte size for either of them.  I did not think of using `cpp_info.defines`, that's a good point actually.

> Also, for the same version of g++, different values of `libcxx` will be completely incompatible binaries, this is why `compiler.libcxx` doesn't have any provision in the `compatibility.py`. The only assumption we have so far for the C++ binaries is compatibility for different `cppstd` standards.

Yes, you are right that different libcxx versions will be incompatible.  That makes sense.  For Fortran this is true of the Fortran standard library that comes with the compiler as well.

> It is different for Conan allowing a setting to not be defined than being the default. If there is a default value for a compiler, for example, gcc 11 uses cppstd=14 by default, then the recommendation is to have to explicitly define it. It could only be undefined for example if some user is using only C packages and not C++, they could leave the `compiler.cppstd` undefined. But that doesn't mean that it will be the compiler default, but that it doesn't apply. This point is important for the `package_id` computation and future compatibility.

OK, I see, thanks for clarifying that, makes sense.  Like I said earlier I'm a bit of a Conan beginner so not sure.  :-)

**Comment by dnwillia** [comment]
> From a metadata perspective, I think "gfortran" as a compiler value makes sense for projects that are exclusively fortran projects. However, my experience (as a C++ developer) has been that it's more likely that a project depends on both C/C++ and fortran, which complicates things - how can we model the metadata for two different compilers per project? should the metadata be tracked at the library level instead of the package level?

Yes, about your first sentence.  I agree.

Regarding the second bit I have this problem too, i.e. a C++ library that links Fortran into it for some numerical computation libraries that I don't want to build as a separate library.  So g++ and gfortran are used to build code into a single library that exposes a C++ interface.

Since there does not seem to be a way to specify that two compilers are used to construct a single package I kind of thought, well, "one thing at a time" right, let's see where this goes.

Alternative settings maybe?  i.e. something like `settings.fortran_compiler` that is separate from `settings.compiler`.

**Comment by dnwillia** [comment]
Ha, I see @memsharded already suggested the alternative settings.  Full credit goes there.

**Comment by dnwillia** [comment]
> What are the situations in which the ABI of fortran binaries/archives needs to be tracked (aside from modelling dependencies between pure fortran libraries)? Is the risk here that if we have two dependencies that are both C++ and Fortran projects, that if they've used a different fortran compiler version that we'll have an ABI mismatch when both dependencies are static?

I've not been able (yet) to actually get the fortran run times to statically link into a C++ library, so assuming the fortran libraries remain an externalized dependency, then yes, this is a risk.  i.e. two C++ libraries depending on Fortran run times linked into the same application must be compiled with the same gfortran compiler and use the same Fortran run time.  I think it's safe as long each C++ package can specify the fortran compiler version used to build the code within.

> To date the bigger concern that I've had is that the version of libgfortran available on a system doesn't match the version that a binary has been compiled with (or doesn't exist, because it's not part of the default gcc toolchain on most linux systems)

Yeah, that is a concern but it's also true of libstdc++ too if I build my application with a different g++ compiler than what is installed 'out of the box' on the Linux system.  In that case I have to redistribute what I compiled against.

**Comment by memsharded** [comment]
> OK, I see, thanks for clarifying that, makes sense. Like I said earlier I'm a bit of a Conan beginner so not sure. :-)

Very interesting conversation, you are already exploring advanced concepts in Conan, so not that a beginner :)


I suggest to try to move forward: I'll check with the team about this. We don't have the resources right now, but at least to agree on the right approach, in case you or the community want to contribute.

**Comment by dnwillia** [comment]
Just bumping this as I've been thinking about it a bit more this week.  I think all that's needed is to at least add the Fortran compiler to the default settings setup for the `compiler` variable and `version`.  The rest of the things I note, such as the real and integer sizing, should probably be handled via options.

**Comment by memsharded** [comment]
Thanks for the feedback.

But what about the issues that were mentioned above? The problem about new entries in the existing ``compiler`` is that it would only work for pure fortran packages.
But there are some packages that contain mixed C/C++ and fortran. In that regard, it would be better to use a ``fcompiler`` entry (you called it ``fortran_compiler`` I think), so recipes can decide to have ``settings = "compiler", "fcompiler", ...``.

Isn't this the case anymore?

**Comment by dnwillia** [comment]
Yes you are right, for mixed Fortran/C++ you would need separate settings for the Fortran and C++ compilers as you mentioned earlier if a C++ library is dependent on both right.  This is probably the main relevant use case for Conan I think, i.e. where Fortran modules have been embedded inside alongside the C++ code to provide some capability.

It can simplify a lot if you don't add anything for the additional options I mentioned: `integer_size`, `real_size`, `underscoring` and `fortran_std`.  For those it's really only relevant to fortran library modules, and it seems reasonable to delegate their necessity out to the Conan ecosystem users.
